File: js/GF_logistics_Plan.js
class Plan {
    constructor(ShownTab, List_length, Target_Value) {
        this._setResourceIncreasingRate();
        this._setList(List_length);
        this.ShownTab = ShownTab;
        this.CurrentValue_MAX = ShownTab.CurrentValue_MAX;
        if (Target_Value === undefined) {
            this.TargetValue_html = this._CorrectTargetValueHtml(); 
            this.TargetValue = this._CorrectTargetValue();
        }
        else {
            this.TargetValue = Target_Value;
            this.TargetValue_half = this.TargetValue.slice();
            for (var i = 0; i < 8; i++) {
                this.TargetValue_half[i] = this.TargetValue_half[i] * 0.5;
            }
        }
        this._Norm_Target = this._getNorm(this.TargetValue);
    }
    _setResourceIncreasingRate() {
        var TotalRate = Input_getTotalGreatSuccessRate(true);
        this.ResourceIncreasingRate = 1 + (TotalRate) / 200;
    }
    _setList(length) {
        this.List = new Array(length);
        for (var i = 0; i < length; i++) {
            this.List[i] = new Array(13);//13为该方案价值
        }
    }
    //取得输入的TargetValue, 并防止资源(或契约)之间之比过大(最大5000倍)
    _CorrectTargetValueHtml() {
        var TargetValue_html = Input_getTarget_Correct();
        var ResourceValue = TargetValue_html.slice(0, 4);
        var ContractValue = TargetValue_html.slice(4, 8);
        ResourceValue = this._CorrectTargetValueHtml_main(ResourceValue);
        ContractValue = this._CorrectTargetValueHtml_main(ContractValue);
        TargetValue_html = ResourceValue.concat(ContractValue);
        Input_setTarget(TargetValue_html);
        return TargetValue_html;
    }
    _CorrectTargetValueHtml_main(Array4) {
        var MaxValue = ArrayMax(Array4);
        var MinValue = Math.round(MaxValue / 5000 * 100) / 100;
        for (var i = 0; i < 4; i++) {
            if (Array4[i] !== 0)
                Array4[i] = Math.max(MinValue, Array4[i]);
        }
        return Array4;
    }
    _CorrectTargetValue() {
        var ResourceValue = this._CorrectResourceValue();
        var ContractValue = this._CorrectContractValue();
        var TargetValue = ResourceValue.concat(ContractValue);
        if (TargetValue.toString() === "0,0,0,0,0,0,0,0") {
            alert(language.JS.TargetValue0_alert);
            throw"--";
        }
        return TargetValue;
    }
    _CorrectResourceValue() {
        var ResourceValue = this.TargetValue_html.slice(0, 4);
        var Resource_CalibrationValue = 100 - Input_getContractWeight();
        if (this._ValuesNotAll0(ResourceValue))
            this._CorrectValue(ResourceValue, Resource_CalibrationValue);
        return ResourceValue;
    }
    _CorrectContractValue() {
        var ContractValue = this.TargetValue_html.slice(4, 8);
        var Contract_CalibrationValue = Input_getContractWeight();
        if (this._ValuesNotAll0(ContractValue))
            this._CorrectValue(ContractValue, Contract_CalibrationValue);
        return ContractValue;
    }
    _ValuesNotAll0(Values) {
        for (var i = 0; i < Values.length; i++) {
            if (Values[i] !== 0)
                return true;
        }
        return false;
    }
    _CorrectValue(Values, CalibrationValue) {
        var CorrectionRate = CalibrationValue / ArrayMax(Values);
        for (var i = 0; i < Values.length; i++) {
            Values[i] *= CorrectionRate;
        }
    }

	_getNorm(vector) {
		var norm;
		var SumOfSquares = 0;
		for (var i = 0; i < vector.length; i++) {
			SumOfSquares += vector[i] * vector[i];
		}
        norm = Math.pow(SumOfSquares, 0.5);
        return norm;
    }
    
    CalculateAndPush_Standardization_And_CalculateMissionsValue(Mission_n1, Mission_n2, Mission_n3, Mission_n4) {
        this._CurrentValue = this.ShownTab.Calculate_Current(Mission_n1, Mission_n2, Mission_n3, Mission_n4);
        if (this._CurrentValue[0] === -1) {
            return;
        }

        var PlanValue = this._calculateValue();
        var Qvalid = this.ShownTab.Qvalid;
        Qvalid[Mission_n1][11] += PlanValue;
        Qvalid[Mission_n2][11] += PlanValue;
        Qvalid[Mission_n3][11] += PlanValue;
        Qvalid[Mission_n4][11] += PlanValue;
        if (!(0 in this.List[this.List.length - 1]))
            this._push_FirstEmptyRow(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
        else
            this._push(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
    }

    CalculateAndPush_Standardization(Mission_n1, Mission_n2, Mission_n3, Mission_n4) {
        this._CurrentValue = this.ShownTab.Calculate_Current(Mission_n1, Mission_n2, Mission_n3, Mission_n4);
        if (this._CurrentValue[0] === -1) {
            return;
        }
        var PlanValue = this._calculateValue();
        if (!(0 in this.List[this.List.length - 1]))
            this._push_FirstEmptyRow(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
        else
            this._push(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
    }
    CalculateAndPush(Mission_n1, Mission_n2, Mission_n3, Mission_n4) {
        this._CurrentValue = this.ShownTab.Calculate_Current(Mission_n1, Mission_n2, Mission_n3, Mission_n4);
        if (this._CurrentValue[0] === -1) {
            return;
        }
        for (var i = 0; i < 8; i++) {
            if (this._CurrentValue[i] < this.TargetValue_half[i])
                return;
        }
        var PlanValue = this._calculateValue_2();
        if (!(0 in this.List[this.List.length - 1]))
            this._push_FirstEmptyRow(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
        else
            this._push(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
    }
    _push_FirstEmptyRow(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue) {
        var row = this.List.length - 1;
        while (row !== 0 && !(0 in this.List[row - 1]))
            row--;
        this._PushIntoThisRow(row, Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
        this._SortListByValue(row, PlanValue);
    }
    _push(Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue) {
        var ListLastRow = this.List.length - 1;
        if (PlanValue <= this.List[ListLastRow][12])
            return;
        this._PushIntoThisRow(ListLastRow, Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue);
        this._SortListByValue(ListLastRow, PlanValue);
    }
    _PushIntoThisRow(RowNumber, Mission_n1, Mission_n2, Mission_n3, Mission_n4, PlanValue) {
        var _CurrentValue = this._CurrentValue;
        this.List[RowNumber] = [Mission_n1, Mission_n2, Mission_n3, Mission_n4, _CurrentValue[0], _CurrentValue[1], _CurrentValue[2], _CurrentValue[3], _CurrentValue[4], _CurrentValue[5], _CurrentValue[6], _CurrentValue[7], PlanValue];
    }
    _SortListByValue(thisrow, PlanValue) {
        for (var i = thisrow - 1; i >= 0; i--) {
            if (PlanValue > this.List[i][12])
                [this.List[i + 1], this.List[i]] = [this.List[i], this.List[i + 1]];
            else
                break;
        }
    }

    _calculateValue() {
        var CurrentValue = [0, 0, 0, 0, 0, 0, 0, 0];
        var TargetValue = this.TargetValue;
        var _CurrentValue = this._CurrentValue;
        if (TargetValue[0] !== 0)
            CurrentValue[0] = _CurrentValue[0];
        if (TargetValue[1] !== 0)
            CurrentValue[1] = _CurrentValue[1];
        if (TargetValue[2] !== 0)
            CurrentValue[2] = _CurrentValue[2];
        if (TargetValue[3] !== 0)
            CurrentValue[3] = _CurrentValue[3];
        if (TargetValue[4] !== 0)
            CurrentValue[4] = _CurrentValue[4] * 500;
        if (TargetValue[5] !== 0)
            CurrentValue[5] = _CurrentValue[5] * 500;
        if (TargetValue[6] !== 0)
            CurrentValue[6] = _CurrentValue[6] * 500;
        if (TargetValue[7] !== 0)
            CurrentValue[7] = _CurrentValue[7] * 500;
        var Norm_Current = this._getNorm(CurrentValue);
        if (Norm_Current === 0)
            return 0;
        var Dot_product = this._getDotProduct(CurrentValue, this.TargetValue);
        var CurrentScalarProjection = Dot_product / this._Norm_Target;
        var COStheta = CurrentScalarProjection / Norm_Current;
        var theta_ = (-0.698131700797732 * COStheta * COStheta - 0.872664625997164) * COStheta + 1.57079632679489;
        var theta = 0 < theta_ ? theta_ : 0;
        var CosineSimilarity_ = 1 - 0.5 * theta;
        var CosineSimilarity = CosineSimilarity_ * CosineSimilarity_;
        return CurrentScalarProjection * CosineSimilarity;
    }
    // 仅适用于此
    _getDotProduct(vector1, vector2) {
        var Dot_product = 0;
        for (var i = 0; i < 8; i++) {//vector1 == vector2 == 8
            Dot_product += (vector1[i] * vector2[i]);
        }
        return Dot_product;
    }

    _calculateValue_2() {
        return Value2(this.TargetValue, this._CurrentValue);
    }

    print(fineTuningExpanded, SortBy = "Ranking") {
        if (!(0 in this.List[0])) {
            var Table = document.getElementById("Plan_Table");
            Table.innerHTML = language.JS.NoPlan;
            RESULT_PLAN = [];
            throw"--";
        }
        var result_plan = [];
        for (var i = 0; i < this.List.length; i++) {
            if (!(0 in this.List[i])) break;
            var one_plan = [];
            one_plan.push(i);
            var MissionsNumber = new Array(4);
            for (var ii = 0; ii < 4; ii++) {
                MissionsNumber[ii] = this.ShownTab.Qvalid[this.List[i][ii]][0];
            }
            MissionsNumber = MissionsNumber.sort(sortStringNumber);
            for (var ii = 0; ii < 4; ii++) {
                one_plan.push(MissionsNumber[ii]);
            }
            for (var ii = 0; ii < 4; ii++) {
                one_plan.push(this.List[i][ii + 4] * this.ResourceIncreasingRate * this.CurrentValue_MAX[ii]);
            }
            for (var ii = 4; ii < 8; ii++) {
                one_plan.push(this.List[i][ii + 4] * this.CurrentValue_MAX[ii]);
            }
            var customPrint = this.ShownTab.PrintTableCustomize(this, i);
            for (var ii = 0; ii < customPrint.length; ii++) {
                one_plan.push(customPrint[ii]);
            }
            result_plan.push(one_plan);
        }
        var Minutes;
        if (is_CalculateByHour())
            Minutes = 60;
        else
            Minutes = this.ShownTab.TotalTime;

        switch(SortBy) {
            case "Ranking":
                RESULT_PLAN_SORT_BY = "Ranking";
                break;
            case "Manp":
                quick_sort_expand_descending(result_plan, 5);
                RESULT_PLAN_SORT_BY = "Manp";
                break;
            case "Ammu":
                quick_sort_expand_descending(result_plan, 6);
                RESULT_PLAN_SORT_BY = "Ammu";
                break;
            case "Rati":
                quick_sort_expand_descending(result_plan, 7);
                RESULT_PLAN_SORT_BY = "Rati";
                break;
            case "Part":
                quick_sort_expand_descending(result_plan, 8);
                RESULT_PLAN_SORT_BY = "Part";
                break;
            case "TPro":
                quick_sort_expand_descending(result_plan, 9);
                RESULT_PLAN_SORT_BY = "TPro";
                break;
            case "Equi":
                quick_sort_expand_descending(result_plan, 10);
                RESULT_PLAN_SORT_BY = "Equi";
                break;
            case "QPro":
                quick_sort_expand_descending(result_plan, 11);
                RESULT_PLAN_SORT_BY = "QPro";
                break;
            case "QRes":
                quick_sort_expand_descending(result_plan, 12);
                RESULT_PLAN_SORT_BY = "QRes";
                break;
        }
        RESULT_PLAN = result_plan;
        print_result_plan(fineTuningExpanded, result_plan, Minutes);
    }
}

function sortStringNumber(a, b) {
    var aa, bb;
    aa = parseInt(a.replace(/[^0-9]/ig,""));
    bb = parseInt(b.replace(/[^0-9]/ig,""));
    return aa - bb;
}

function print_result_plan(fineTuningExpanded, result_plan, Minutes) {
    //保存Plan Table横向位置, 用于点击排序后或微调后恢复位置
    var result_plan_table_scrollLeft = 0;
    if (document.getElementById("result_plan_table") !== null)
        result_plan_table_scrollLeft = document.getElementById("result_plan_table").scrollLeft;

    //保存打印前聚焦元素, 用于微调后恢复聚焦元素
    var focusedElementID = document.activeElement.id;

    var Table = document.getElementById("Plan_Table");
    var tab = getHTMLFineTuningTool();
    tab += '<div class="table-responsive" id="result_plan_table">';
    tab += '<table class="table table-striped table-hover table-responsive text-nowrap" style="margin-bottom: 0px; cursor: default;">';
    var ShownTab = getShownTab();
    tab += (ShownTab.PrintPlanTableTitle() + '<tbody>');
    var is_selected;
    var selectedMissions;
    if (MISSION_TABLE_SELECT.length === 4) {
        is_selected = true;
        selectedMissions = MISSION_TABLE_SELECT.slice();
        selectedMissions.sort(sortStringNumber);
    }
    else
        is_selected = false;
    for (var i = 0; i < result_plan.length; i++) {
        if (is_selected) {
            if (result_plan[i][1] === selectedMissions[0] && result_plan[i][2] === selectedMissions[1] && result_plan[i][3] === selectedMissions[2] && result_plan[i][4] === selectedMissions[3])
                tab += "<tr tabindex='0' id='print_result_plan_tr_" + i + "' class='success'>";
            else
                tab += "<tr tabindex='0' id='print_result_plan_tr_" + i + "'>";
        }
        else
            tab += "<tr tabindex='0' id='print_result_plan_tr_" + i + "'>";
        for (var ii = 0; ii < 4; ii++) {
            tab += "<td style='text-align: center; width: 5%;'>";
            tab += "" + result_plan[i][ii + 1];
            tab += "</td>";
        }
        for (var ii = 4; ii < 8; ii++) {
            tab += "<td style='text-align: center; width:" + ShownTab.PlanTableResourceAndContractWidth + "'>";
            tab += "" + Math.round(result_plan[i][ii + 1] * Minutes * 10) / 10;
            tab += "</td>";
        }
        for (var ii = 8; ii < 12; ii++) {
            tab += "<td style='text-align: center; width:" + ShownTab.PlanTableResourceAndContractWidth + ";'>";
            tab += "" + Math.round(result_plan[i][ii + 1] * Minutes * 100) / 100;
            tab += "</td>";
        }
        var one_plan_length = result_plan[0].length;
        for (var ii = 13; ii < one_plan_length; ii++) {
            tab += "<td style='text-align: center; width:" + ShownTab.PlanTableResourceAndContractWidth + ";'>";
            tab += result_plan[i][ii];
            tab += "</td>";
        }
        tab += "</tr>";
    }
    tab += '</tbody>';
    Table.innerHTML = tab;
    if (fineTuningExpanded) {
        document.getElementById("FineTuningTool").style.transition = "none";
        $("#FineTuningTool").collapse("show");
    }
    if (IsMobile())
        document.getElementById("FineTuningTool").style.transition = "none";
    document.getElementById("start_sorting_html").style.display = "none";
    if (focusedElementID !== "")
        document.getElementById(focusedElementID).focus();
    document.getElementById("result_plan_table").scrollLeft = result_plan_table_scrollLeft;
}

File: tools/Generate Training Data/GF_logistics_Plan.cpp
#include"GF_logistics.h"
#include"Tab.h"
#define PI acos(-1)

extern array<double, 7> TargetValue_0;
extern Tab* ShownTab;

extern array<int, 48> MissionsNumber_times;
extern int test;
extern int test_2;
extern fstream txt_Ok;
extern fstream txt_noOk;

class Plan {
    public:
		Plan(int length);
		void CalculateAndPush(int* MissionsNumber);
		void printMissionsNumber_times();
		void print();

    private:
		vector<array<double, 14>> List;
        array<double, 7> TargetValue;
        int List_length;

        double Norm_Target;
        int _MissionsNumber[4];
        array<double, 7> _CurrentValue;
        double _PlanValue_1;
		double _PlanValue_2;

        double getDotProduct(array<double, 7>vector1, array<double, 7>vector2);
        double calculateValue_1();
		double Value_0(double Target, double Current);
		double calculateValue_2();
        bool thisPlanIsBetterThan(int row);
        bool eachCurrentValueIsBigger(int row);

        array<double, 7> CorrectTargetValue() {
            array<double, 4> ResourceValue = CorrectResourceValue();
            array<double, 3> ContractValue = CorrectContractValue();
            array<double, 7> Total;
            for (int i = 0; i < 4; i++) {
                Total[i] = ResourceValue[i];
            }
            for (int i = 4; i < 7; i++) {
                Total[i] = ContractValue[i - 4];
            }
            return Total;
        }
        array<double, 4> CorrectResourceValue() {
            array<double, 4> ResourceValue;
            double Resource_CalibrationValue = 1000;
            for (int i = 0; i < 4; i++) {
                ResourceValue[i] = TargetValue_0[i];
            }
            if (ValuesNotAll0(ResourceValue)) {
                ResourceValue = CorrectValue(ResourceValue, Resource_CalibrationValue);
            }
            return ResourceValue;
        }
        array<double, 3> CorrectContractValue() {
            array<double, 3> ContractValue;
            double Contract_CalibrationValue = 1000;
            for (int i = 0; i < 3; i++) {
                ContractValue[i] = TargetValue_0[i + 4];
            }
            if (ValuesNotAll0(ContractValue)) {
                ContractValue = CorrectValue(ContractValue, Contract_CalibrationValue);
            }
            return ContractValue;
        }
        template<typename T> bool ValuesNotAll0(T Values) {
            for (int i = 0; i < Values.size(); i++) {
                if (Values[i] != 0) return true;
            }
            return false;
        }
        template<typename T> T CorrectValue(const T Values, double CalibrationValue) {
            T Values_0;
            double CorrectionRate = CalibrationValue / ArrayMax(Values);
            for (int i = 0; i < Values.size(); i++) {
                Values_0[i] = CorrectionRate * Values[i];
            }
            return Values_0;
        }
        template<typename T> double ArrayMax(T Arr) {
            double maxValue = 0;
            for (int i = 0; i < Arr.size(); i++) {
                maxValue = max(maxValue, Arr[i]);
            }
            return maxValue;
        }

        template<typename T> double getNorm(T Value_vector) {
            double norm;
            double SumOfSquares = 0;
            for (int i = 0; i < Value_vector.size(); i++) {
                SumOfSquares += pow(Value_vector[i], 2);
            }
            norm = pow(SumOfSquares, 0.5);
            return norm;
        }

        void push_FirstEmptyrow() {
            int row = List_length - 1;
            while (row != 0 && List[row - 1][2] == 0 && List[row - 1][3] == 0) row--;
            PushIntoThisRow(row);
            SortListByValue(row);
        }
        void push() {
            if (!thisPlanIsBetterThan(List_length - 1)) return;
            PushIntoThisRow(List_length - 1);
            SortListByValue(List_length - 1);
        }
        void PushIntoThisRow(int RowNumber) {
            for (int i = 0; i < 4; i++) {
                List[RowNumber][i] = _MissionsNumber[i];
            }
            for (int i = 0; i < 7; i++) {
                List[RowNumber][i + 4] = _CurrentValue[i];
            }
            List[RowNumber][11] = _PlanValue_1;
			List[RowNumber][12] = _PlanValue_2;
        }
        void SortListByValue(int thisrow) {
            for (int i = thisrow - 1; i >= 0; i--) {
                if (thisPlanIsBetterThan(i)) {
                    ExchangeTheseTwoRows(i);
                }
                else break;
            }
        }
        void ExchangeTheseTwoRows(int RowNumber) {
            for (int i = 0; i < 14; i++) {
                List[RowNumber + 1][i] = List[RowNumber][i];
            }
            PushIntoThisRow(RowNumber);
        }
};

Plan::Plan(int length) {
	List_length = length;
	array<double, 14> row = { 0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
	for (int i = 0; i < List_length; i++) {
		//List.insert(List.begin(), 1, row);
		List.push_back(row);
	}
	TargetValue = CorrectTargetValue();
	Norm_Target = getNorm(TargetValue);
}

void Plan::CalculateAndPush(int* MissionsNumber) {
	_MissionsNumber[0] = MissionsNumber[0];
	_MissionsNumber[1] = MissionsNumber[1];
	_MissionsNumber[2] = MissionsNumber[2];
	_MissionsNumber[3] = MissionsNumber[3];
	_CurrentValue = ShownTab->Calculate_Current(MissionsNumber);
	_PlanValue_1 = calculateValue_1();
	_PlanValue_2 = calculateValue_2();
	if (List[List_length - 1][2] == 0 && List[List_length - 1][3] == 0) {
		push_FirstEmptyrow();
	}
	else push();
}

double Plan::calculateValue_1() {
	test++;
    _CurrentValue[4] *= 500;
    _CurrentValue[5] *= 500;
    _CurrentValue[6] *= 500;
    double Norm_Current = getNorm(_CurrentValue);
    double Dot_product = getDotProduct(_CurrentValue, TargetValue);
    double CurrentScalarProjection = Dot_product / Norm_Target;
    double COStheta = CurrentScalarProjection / Norm_Current;
    double theta = acos(COStheta);
    double CosineSimilarity_0 = 1 - 2 * theta / PI;
    double CosineSimilarity = pow(CosineSimilarity_0, 2);
    _CurrentValue[4] /= 500;
    _CurrentValue[5] /= 500;
    _CurrentValue[6] /= 500;
    return CurrentScalarProjection * CosineSimilarity;
}
double Plan::getDotProduct(array<double, 7>vector1, array<double, 7>vector2) {
    double Dot_product = 0;
    for (int i = 0; i < 7; i++) {
        Dot_product += (vector1[i] * vector2[i]);
    }
    return Dot_product;
}

double Plan::calculateValue_2() {
	return Value_0(TargetValue[0], _CurrentValue[0]) + Value_0(TargetValue[1], _CurrentValue[1]) +
		Value_0(TargetValue[2], _CurrentValue[2]) + Value_0(TargetValue[3], _CurrentValue[3]);
}
double Plan::Value_0(double Target, double Current) {
	if (Target == 0) return 0;
	if (Target > Current) {//Y=5.5*x^3+4.5*x
		return (Target - Current) * (5.5 * pow((Target - Current) / Target, 3) + 4.5 * (Target - Current) / Target);
	}
	else {//Y=-0.04*x^3 -0.01*x
		return (Target - Current) * (-0.04 * pow((Target - Current) / Target, 3) - 0.01 * (Target - Current) / Target);
	}
}

bool Plan::thisPlanIsBetterThan(int number) {
	test_2++;
    /*if (eachCurrentValueIsBigger(number)) return true;
    else {
        if (_PlanValue_1 > List[number][11]) return true;
        else return false;
    }*/
	if ((_PlanValue_1 > List[number][11] && _PlanValue_2 < List[number][12])|| (_PlanValue_1 <= List[number][11] && _PlanValue_2 >= List[number][12])) {
		txt_Ok << TargetValue[0] << ",";
		txt_Ok << TargetValue[1] << ",";
		txt_Ok << TargetValue[2] << ",";
		txt_Ok << TargetValue[3] << ",";
		txt_Ok << TargetValue[4] << ",";
		txt_Ok << TargetValue[5] << ",";
		txt_Ok << TargetValue[6] << ",";
		txt_Ok << _CurrentValue[0] << ",";
		txt_Ok << _CurrentValue[1] << ",";
		txt_Ok << _CurrentValue[2] << ",";
		txt_Ok << _CurrentValue[3] << ",";
		txt_Ok << _CurrentValue[4] << ",";
		txt_Ok << _CurrentValue[5] << ",";
		txt_Ok << _CurrentValue[6] << ",";
		txt_Ok << List[number][4] << ",";
		txt_Ok << List[number][5] << ",";
		txt_Ok << List[number][6] << ",";
		txt_Ok << List[number][7] << ",";
		txt_Ok << List[number][8] << ",";
		txt_Ok << List[number][9] << ",";
		txt_Ok << List[number][10] << ",";
		if (_PlanValue_1 > List[number][11] && _PlanValue_2 < List[number][12]) {
			txt_Ok << 0 << endl;
		}
		else {
			txt_Ok << 1 << endl;
		}
		
	}
	else {
		txt_noOk << TargetValue[0] << ",";
		txt_noOk << TargetValue[1] << ",";
		txt_noOk << TargetValue[2] << ",";
		txt_noOk << TargetValue[3] << ",";
		txt_noOk << TargetValue[4] << ",";
		txt_noOk << TargetValue[5] << ",";
		txt_noOk << TargetValue[6] << ",";
		txt_noOk << _CurrentValue[0] << ",";
		txt_noOk << _CurrentValue[1] << ",";
		txt_noOk << _CurrentValue[2] << ",";
		txt_noOk << _CurrentValue[3] << ",";
		txt_noOk << _CurrentValue[4] << ",";
		txt_noOk << _CurrentValue[5] << ",";
		txt_noOk << _CurrentValue[6] << ",";
		txt_noOk << List[number][4] << ",";
		txt_noOk << List[number][5] << ",";
		txt_noOk << List[number][6] << ",";
		txt_noOk << List[number][7] << ",";
		txt_noOk << List[number][8] << ",";
		txt_noOk << List[number][9] << ",";
		txt_noOk << List[number][10] << ",";
		txt_noOk << endl;
	}

	if (_PlanValue_1 > List[number][11]) return true;
	else return false;
}
bool Plan::eachCurrentValueIsBigger(int number) {
    for (int i = 0; i < 7; i++) {
        if (_CurrentValue[i] < List[number][i + 4]) return false;
    }
    return true;
}

void Plan::printMissionsNumber_times() {
	for (int i = 0; i < List_length; i++) {
		for (int ii = 0; ii < 4; ii++) {
			MissionsNumber_times[List[i][ii]]++;
		}
	}
}

void Plan::print() {

}